packaging/markers.py

```diff
@@ -288,6 +322,15 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f"<Marker('{self}')>"
 
+    def __hash__(self) -> int:
+        return hash((self.__class__.__name__, str(self)))
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Marker):
+            return NotImplemented
+
+        return _flatten_marker(self._markers) == _flatten_marker(other._markers)
```

Comment:
An alternative to flattening the `_markers` list here would be doing it directly on the `__init__` method.

Reply:
As in pre-compute the string? Is the string used enough to warrant the forced cost of doing that?

Reply:
Not necessarily the string. The key point here seems to be flattening parenthesized groups with a single element (or the data structure equivalent to the parsing of those).

The "flattening" seems to be one of the central keys of the string conversion too.

We could do this flattening in the `__init__` function without pre-computing the string, and that would facilitate both `__eq__` and `__str__`.

Reply:
My vote is to rely on the string representation for simplicity, else we are duplicating algorithms for walking the markers. If it turns out performance is a problem we can cache it, but we can wait on that until that actually becomes a problem.

Reply:
Yeah, I think that definitely is the simplest approach. I have updated the implementation accordingly.

(To be honest my first impulse was to do a string comparison, but I refrained from adopting that by overthinking about a [previous comment](https://github.com/pypa/packaging/pull/499#discussion_r788124952), in a different context, about the cost of the string conversion).